src/core/lib/security/transport/security_handshaker.c

```diff
@@ -223,17 +228,23 @@ static grpc_error *on_handshake_next_done_locked(
         GRPC_ERROR_CREATE_FROM_STATIC_STRING("Handshake failed"), result);
   }
   // Send data to peer.
-  grpc_slice to_send = grpc_slice_from_copied_buffer(
-      (const char *)bytes_to_send, bytes_to_send_size);
-  grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
-  grpc_slice_buffer_add(&h->outgoing, to_send);
-  grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
-                      &h->on_handshake_data_sent_to_peer);
-  // If handshake has completed, check peer and so on.
+  if (bytes_to_send_size > 0) {
+    grpc_slice to_send = grpc_slice_from_copied_buffer(
+        (const char *)bytes_to_send, bytes_to_send_size);
+    grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
+    grpc_slice_buffer_add(&h->outgoing, to_send);
+    grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
+                        &h->on_handshake_data_sent_to_peer);
+  }
+  // If handshake has completed, check peer and so on. Otherwise, need to read
+  // more data from the peer.
   if (handshaker_result != NULL) {
     GPR_ASSERT(h->handshaker_result == NULL);
     h->handshaker_result = handshaker_result;
     error = check_peer_locked(exec_ctx, h);
+  } else {
+    grpc_endpoint_read(exec_ctx, h->args->endpoint, h->args->read_buffer,
```

Comment:
I agree doing both read and write at the same time is not good (although code is cleaner).

What you propose is problematic. There is a situation where handshake completes (handshake_result is not NULL) and there are some final bytes to send to peer . If we make line 241 mutually exclusive, we send the bytes (without setting the h->handshaker_result). In `on_handshake_data_send_to_peer()`, we need to wait for grpc read, but we will never get any data back. 

Also, if remove condition in `on_handshake_data_send_to_peer()` and always do read. The read could be blocked, because handshake finished and there is nothing to read.

Basically, there are two corner cases security handshaker need to handle
1. There is nothing to send to peer, but need to read from peer
2. There are some bytes (i.e., finishing bytes) to send to peer, but no need to read from peer.

I am thinking about two possibilities here:
A. rollback last commit of this pull request, it should work fine. Doing a grpc_endpoint_write on an empty buffer is OK, as security handshaker already did it today.
B. In `on_handshake_data_sent_to_peer()`, do a conditional read. In `on_handshake_next_done_locked`, have 
```
if (bytes_to_send_size > 0) {
  // there are something to send
  ... do write ...
} else if (handshaker_result == NULL) {
  // nothing to send, but need to read
  ... do read ...
}

// handshake completes is regardless of write. There could be final write and handshake completes
if (handshaker_result != NULL) {
  check_peer_locked();
} 

```